linq4j/src/main/java/org/apache/calcite/linq4j/EnumerableDefaults.java

```diff
@@ -1063,22 +1084,25 @@ public void close() {
       final Enumerable<TSource> outer, final Enumerable<TInner> inner,
       final Function1<TSource, TKey> outerKeySelector,
       final Function1<TInner, TKey> innerKeySelector,
+      final Predicate2<TSource, TInner> predicate,
       final Function2<TSource, TInner, TResult> resultSelector,
       final EqualityComparer<TKey> comparer, final boolean generateNullsOnLeft,
       final boolean generateNullsOnRight) {
     return new AbstractEnumerable<TResult>() {
       public Enumerator<TResult> enumerator() {
+        final List<TInner> rightList = inner.toList();
         final Lookup<TKey, TInner> innerLookup =
             comparer == null
-                ? inner.toLookup(innerKeySelector)
-                : inner.toLookup(innerKeySelector, comparer);
+                ? Linq4j.asEnumerable(rightList).toLookup(innerKeySelector)
```

Comment:
Maybe I'm missing something, but it seems that we have converted an enumerable (`inner`) into a list (`rightList`), and then back to enumerable (`Linq4j.asEnumerable(rightList)`). Wouldn't it be simpler to keep this line and the next one unchanged (using `inner`)?

Reply:
@rubenada , we want to get the rightUnmatched, so the inner would be figured out at first.
`        final List<TInner> rightList = inner.toList();
`
 Converting the rightList back to enumerable would't increase the overhead,
because `Linq4j.asEnumerable(rightList)`  that will produce a ListEnumerable.
If we keep the original inner, when doing `toLookup`, this inner will be figured out again , it's unnecessary.

Reply:
but if the join type is not right outer,  we don't need to figure out the rightUnmatched. So we can use the inner to do `toLookup`.

Reply:
@LaiZhou, ok, I understand, thanks for the explanation.
Small detail, maybe instead of "right" (rightList / rightUnmatched) we should use "inner" instead (to be consistent with the rest of variables in the method)?

Reply:
@rubenada , I update the PR to make the variables more clear.
`  final Enumerable<TInner> innerToLookUp = generateNullsOnLeft
            ? Linq4j.asEnumerable(inner.toList())
            : inner;
        final Lookup<TKey, TInner> innerLookup =
            comparer == null
                ? innerToLookUp.toLookup(innerKeySelector)
                : innerToLookUp
                    .toLookup(innerKeySelector, comparer);`

`List<TInner> rightUnmatched =
              generateNullsOnLeft
                  ? new ArrayList<>(innerToLookUp.toList())
                  : null;`